text/0000-stable-wakers.md

```diff
@@ -0,0 +1,196 @@
+	other_adapter: unsafe extern "C" fn(unsafe fn(*const ()), *const ()),
+}
+impl RawWakerVTable {
+    pub const fn new(
+        clone: unsafe fn(*const ()) -> RawWaker,
+        wake: unsafe fn(*const ()),
+        wake_by_ref: unsafe fn(*const ()),
+        drop: unsafe fn(*const ()),
+    ) -> Self {
+		unsafe extern "C" fn clone_adapter(clone: unsafe fn(*const ()) -> RawWaker, data: *const ()) -> RawWaker {
+			clone(data)
+		}
+		unsafe extern "C" fn other_adapter(other: unsafe fn(*const ()), data: *const ()) {
+			other(data)
+		}
+        Self { clone, wake, wake_by_ref, drop, clone_adapter, other_adapter }
+    }
+}
+impl Waker {
+	pub fn wake(self) {
+        let other_adapter = self.waker.vtable.other_adapter;
+        let wake = self.waker.vtable.wake;
+        let data = self.waker.data;
+        // Don't call `drop` -- the waker will be consumed by `wake`.
+        crate::mem::forget(self);
+        // SAFETY: This is safe because `Waker::from_raw` is the only way
+        // to initialize `wake` and `data` requiring the user to acknowledge
+        // that the contract of `RawWaker` is upheld.
+
+        // This is also FFI-safe because `other_adapter` adapts the calling convention.
+        unsafe { (other_adapter)(wake, data) };
+    }
+    // Similar implementation changes to `wake_by_ref` and `drop`
+}
+impl Clone for Waker {
+    #[inline]
+    fn clone(&self) -> Self {
+        Waker {
+            // SAFETY: This is safe because `Waker::from_raw` is the only way
+            // to initialize `clone` and `data` requiring the user to acknowledge
+            // that the contract of [`RawWaker`] is upheld.
+            waker: unsafe { (self.waker.vtable.clone_adapter)(clone, self.waker.data) },
+        }
+    }
+}
+```
+
+However, this adds a level of indirection to function calls. Providing a `RawWakerVTable::new_with_c_abi` constructor, and making it a stable niche optimized sum type like so would allow removing that indirection when the vtable is constructed from `extern "C" fn`s:
+```rust
+#[repr(C)]
+struct OldRawWakerVTable {
+    clone: unsafe fn(*const ()) -> RawWaker,
+    wake: unsafe fn(*const ()),
+    wake_by_ref: unsafe fn(*const ()),
+    drop: unsafe fn(*const ()),
+	clone_adapter: Option<unsafe extern "C" fn(unsafe fn(*const ()) -> RawWaker, *const ()) -> RawWaker>,
+	other_adapter: Option<unsafe extern "C" fn(unsafe fn(*const ()), *const ())>,
+    padding: [*const (); 2] // Reserved space for future changes to the v-table, the need for it and amount are debatable.
+}
+#[repr(C)]
+struct NewRawWakerVTable {
+    clone: unsafe extern "C" fn(*const ()) -> RawWaker,
+    wake: unsafe extern "C" fn(*const ()),
+    wake_by_ref: unsafe extern "C" fn(*const ()),
+    drop: unsafe extern "C" fn(*const ()),
+    padding: [*const (); 4] // Reserved space for future changes to the v-table, the need for it and amount are debatable.
```

Comment:
I'd suggest to not stabilize the size of the vtable at all, so that adding more fields at the end is not considered breaking.

Reply:
I agree, considering the v-tables are always passed by reference, their size doesn't matter: if later versions of the v-table are created, they would first check that the v-table is in its later variant (presumably by checking if one of the common function pointers between the 2 proposed v-tables is null) before attempting to access any further field.